source/h2_stream.c

```diff
@@ -975,6 +1009,26 @@ struct aws_h2err aws_h2_stream_on_decoder_end_stream(struct aws_h2_stream *strea
      * an actual frame type. It's a flag on DATA or HEADERS frames, and we
      * already checked the legality of those frames in their respective callbacks. */
 
+    if (stream->thread_data.check_incoming_content_length &&
+        stream->thread_data.incoming_data_length != stream->thread_data.incoming_content_length) {
+        /**
+         * RFC-9113 8.1.1:
+         * A request or response is also malformed if the value of a content-length header field does not equal the
+         * sum of the DATA frame payload lengths that form the content, unless the message is defined as having no
+         * content.
+         *
+         * Clients MUST NOT accept a malformed response.
+         */
+        AWS_H2_STREAM_LOGF(
+            ERROR,
+            stream,
+            "Total received data payload=%" PRIu64 " does not match the received content-length header, which=%" PRIi64
+            ". Closing malformed stream",
+            stream->thread_data.incoming_data_length,
+            stream->thread_data.incoming_content_length);
+        return s_send_rst_and_close_stream(stream, aws_h2err_from_h2_code(AWS_HTTP2_ERR_PROTOCOL_ERROR));
```

Comment:
double check that this is the right way to handle malformed msg errors.

I forget the nuances, but the decoder needs to keep processing the HEADER frames because HPACK

Reply:
> Malformed requests or responses that are detected MUST be treated as a [stream error](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#StreamErrorHandler) ([Section 5.4.2](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#StreamErrorHandler)) of type [PROTOCOL_ERROR](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#PROTOCOL_ERROR).

We did it right